crates/api/src/fetch.rs

```diff
@@ -92,40 +97,47 @@ pub type DynFetchFactory = Arc<dyn FetchFactory>;
 
 #[cfg(test)]
 mod test {
-    use crate::MetaOp;
+    use crate::{id::Id, MetaOp};
 
     use super::*;
+    use bytes::BytesMut;
+    use k2_fetch_proto::FetchMessageType;
     use prost::Message;
 
     #[test]
-    fn happy_op_ids_encode_decode() {
+    fn happy_request_encode_decode() {
         let op_id_1 = OpId::from(bytes::Bytes::from_static(b"some_op_id"));
         let op_id_2 = OpId::from(bytes::Bytes::from_static(b"another_op_id"));
         let op_id_vec = vec![op_id_1, op_id_2];
-        let op_ids = OpIds::from(op_id_vec.clone());
+        let op_ids = Request::from(op_id_vec.clone());
 
         let op_ids_enc = op_ids.encode_to_vec();
-        let op_ids_dec = OpIds::decode(op_ids_enc.as_slice()).unwrap();
+        let op_ids_dec = Request::decode(op_ids_enc.as_slice()).unwrap();
         let op_ids_dec_vec = Vec::from(op_ids_dec.clone());
 
         assert_eq!(op_ids, op_ids_dec);
         assert_eq!(op_id_vec, op_ids_dec_vec);
     }
 
     #[test]
-    fn bytes_from_op_ids() {
-        let op_id_1 = OpId::from(bytes::Bytes::from_static(b"some_op_id"));
-        let op_id_2 = OpId::from(bytes::Bytes::from_static(b"another_op_id"));
-        let op_ids = vec![op_id_1, op_id_2];
-
-        let bytes = serialize_op_ids(op_ids.clone());
-        let op_ids_deserialized = deserialize_op_ids(bytes.clone()).unwrap();
-
-        assert_eq!(op_ids_deserialized, op_ids);
+    fn unhappy_decode() {
+        let op_1 = MetaOp {
+            op_id: OpId::from(bytes::Bytes::from_static(b"some_op_id")),
+            op_data: vec![0],
+        };
+        println!("op id {}", op_1.op_id);
+        let ops_enc = Response::from(vec![op_1]).encode_to_vec();
+        let ops_dec = Request::decode(ops_enc.as_slice()).unwrap();
+        let op_ids_dec_vec = ops_dec
+            .op_ids
+            .into_iter()
+            .map(Into::into)
+            .collect::<Vec<OpId>>();
+        println!("op ids dec {op_ids_dec_vec:?}");
```

Comment:
This is an example of where decoding to the wrong type doesn't fail and yields an empty vector. I'll delete this before merging the PR.

Reply:
I think this is expected behavior though right? It's why things like this exist, to help coordinate schemas across applications https://buf.build/product/bsr

At least that's why I've used the Confluent schema registry with Avro when trying to use binary encoding over Kafka.

Reply:
What is a schema used for here?

It is probably nothing we prevent though. If the op is just a byte array, and a response is a vector of such ops, and an op id is a byte array, and the request is a vector of such op ids, then deserializing ops into op ids results in the op ids being the entire ops. As there's no length to op ids, they may well be byte arrays of the length of the whole op.

Reply:
In the cases I know about, you use a serializer that knows about the schema types and the schema registry. It's all slightly magical but the outcome is just a "type" and a "version" that tell you exactly what schema, schema version and message kind are in use. Literally just some extra bytes on the start of the message to avoid exactly what you're running into.